### PR TITLE
Disable app token creation for impersonated people, ref #15539

### PR DESCRIPTION
**Pull request by GretaD**
ref #15539
I created a new method for getter and setter, so impersonate app would not set and get the variable directly, but use this method.

### CODE REVIEW

File: lib/public/IUserSession.php [hunk 1]
```diff
@@ -80,4 +80,17 @@ public function getUser();
 	 * @since 8.0.0
 	 */
 	public function isLoggedIn();
+
+	/**
+	 * get getImpersonatingUserID
+	 * @return string | null
```
Comment (nickvergessen) [inline lib/public/IUserSession.php]:
```suggestion
	 * @return string|null
```

File: lib/public/IUserSession.php [hunk 2]
```diff
@@ -80,4 +80,17 @@ public function getUser();
...
+	 * @since 17.0.0
+	 */
+	public function getImpersonatingUserID();
```
Comment (nickvergessen) [inline lib/public/IUserSession.php]:
```suggestion
	public function getImpersonatingUserID(): ?string;
```
(or is that only with the next php version?)

File: lib/public/IUserSession.php [hunk 3]
```diff
@@ -80,4 +80,17 @@ public function getUser();
...
+
+	/**
+	 * set setImpersonatingUserID
+	 * @since 17.0.0
+	 */
+	public function setImpersonatingUserID(bool $useCurrentUser = true);
```
Comment (nickvergessen) [inline lib/public/IUserSession.php]:
```suggestion
	public function setImpersonatingUserID(bool $useCurrentUser = true): void;
```

**Comment by nickvergessen** [comment]
This should also be checked on the actual endpoint that is being called.

File: lib/public/IUserSession.php [hunk 4]
```diff
@@ -80,4 +80,17 @@ public function getUser();
 	 * @since 8.0.0
 	 */
 	public function isLoggedIn();
+
+	/**
+	 * get getImpersonatingUserID
+	 * @return string|null
+	 * @since 17.0.0
+	 */
+	public function getImpersonatingUserID();
```
Comment (nickvergessen) [inline lib/public/IUserSession.php]:
```suggestion
	public function getImpersonatingUserID(): ?string {
```

can be applied

**Review by nickvergessen — APPROVED** [review]

**Review by ChristophWurst — APPROVED** [review]

**Review by nickvergessen — APPROVED** [review]

**Comment by nickvergessen** [comment]
@GretaD time to fixup and rebase

File: lib/public/IUserSession.php [hunk 5]
```diff
@@ -80,4 +82,19 @@ public function getUser();
...
 	public function isLoggedIn();
-}
+
+	/**
+	 * get getImpersonatingUserID
+	 *
+	 * @return string|null
+	 * @since 17.0.0
+	 */
+	public function getImpersonatingUserID(): ?string;
```
Comment (nickvergessen) [inline lib/public/IUserSession.php]: PHP Fatal error: Class Mock_IUserSession_84aca509 contains 2 abstract methods and must therefore be declared abstract or implement the remaining methods (OCP\IUserSession::getImpersonatingUserID, OCP\IUserSession::setImpersonatingUserID) in phar:///usr/local/bin/phpunit/phpunit-mock-objects/Generator.php(264) : eval()'d code on line 

seems like there are some usages in unit tests which are not working as expected

**Comment by blizzz** [comment]
There are still some errors from testing, e.g.: 

```
1) Test\Settings\Controller\AuthSettingsControllerTest::testCreate
TypeError: Argument 7 passed to OC\Settings\Controller\AuthSettingsController::__construct() must implement interface OCP\IUserSession, instance of Mock_IManager_2651a371 given, called in /drone/src/tests/Settings/Controller/AuthSettingsControllerTest.php on line 81
/drone/src/settings/Controller/AuthSettingsController.php:89
/drone/src/tests/Settings/Controller/AuthSettingsControllerTest.php:81 
```

and the webpack build complains about uncomitted changes.

@GretaD do you want to give it a look?

**Comment by blizzz** [comment]
Now I think you have to rebuild the JS part (`npm run build`) and commit the related changes.

**Comment by rullzer** [comment]
This would then require also additions to the impersonate app :wink:

**Comment by rullzer** [comment]
I updated the phpdoc and rebased. Will emrge after

**Pull request merged**